packages/cmr-client/src/EarthdataToken.ts

```diff
@@ -0,0 +1,183 @@
+    expiration_date?: string
+  }
+}>;
+
+export interface EarthdataTokenParams {
+  username: string,
+  password: string,
+  edlEnv: string,
+  token?: string,
+}
+
+/**
+ * A class to simplify requests for the Earthdata Login token used for CMR
+ *
+ * @typicalname earthdataToken
+ *
+ * @example
+ * const { EarthdataToken } = require('@cumulus/cmr-client');
+ *
+ * const earthdataToken = new EarthdataToken({
+ *  "username": "my-username",
+ *  "password": "my-password",
+ *  "edlEnv": "my-cmr-environment",
+ *  "token" : "my-token"
+ * });
+ */
+export class EarthdataToken {
+  username: string;
+  password: string;
+  edlEnv: string;
+  token?: string;
+
+  /**
+  * The constructor for the EarthdataToken class
+  *
+  * @param {string} params.username - Earthdata username
+  * @param {string} params.password - Earthdata password
+  * @param {string} params.edlEnv - the CMR environment of the user
+  * @param {string} params.token- the EarthdataLogin token, undefined if user doesn't have one
+  *
+  * @example
+  *
+  * {
+  *  "username": "janedoe",
+  *  "password": "password",
+  *  "edlEnv": "UAT",
+  *  "token" : "1782hg134bsd71"
+  *
+  * }
+  */
+
+  constructor(params: EarthdataTokenParams) {
+    this.username = params.username;
+    this.password = params.password;
+    this.edlEnv = params.edlEnv;
+    this.token = params.token;
+  }
+
+  getEDLurl(
+  ) {
+    switch (this.edlEnv) {
+      case 'PROD':
+      case 'OPS':
+        return 'https://urs.earthdata.nasa.gov';
+      case 'UAT':
+        return 'https://uat.urs.earthdata.nasa.gov';
+      case 'SIT':
+        return 'https://sit.urs.earthdata.nasa.gov';
+      default:
+        return 'https://sit.urs.earthdata.nasa.gov';
+    }
+  }
+
+  async getEDLToken(): Promise<string> {
+    if (!this.token) {
+      const buff = Buffer.from(`${this.username + ':' + this.password}`).toString('base64');
+      const url = this.getEDLurl();
+      // response: get a token from the Earthdata login endpoint using credentials if exists
+      let response: EarthdataGetTokenResponse;
+      try {
+        response = await got.get(`${url}/api/users/tokens`,
+          {
+            headers: {
+              Authorization: `Basic ${buff}`,
+            },
+          }).json();
+      } catch (error) {
+        const statusCode = get(error, 'response.statusCode', error.code);
+        const statusMessage = get(error, 'response.statusMessage', error.message);
+        let errorMessage = `Authentication error: Invalid Credentials, Authentication with Earthdata Login failed, statusCode: ${statusCode}, statusMessage: ${statusMessage}`;
+        const responseError = get(error, 'response.body.errors');
+        if (responseError) {
+          errorMessage = `${errorMessage}, CMR error message: ${JSON.stringify(responseError)}`;
+        }
+
+        throw new Error(errorMessage);
+      }
+      if (Object.keys(response).length === 0) {
+        return this.createEDLToken();
```

Comment:
To make code more clear, suggest to have a separate method to getToken.  We can have methods:
getToken (call retrieveToken and createToken)
retrieveToken (retrieve existing valid token if any)
createToken
revokeToken

Remove `EDL` from method name since they are methods in class EarthdataLogin

Reply:
Please add doc to class and methods

Reply:
We want to make the code more clearer and maintainable.  The current getToken method retrieves the token from EDL, and also does this following and makes the logic hard to follow.
if (Object.keys(response).length === 0) {
        return this.createEDLToken();

We can do something for getToken like:
let token = await this.retrieveToken();
if (token === undefined) {
token = await this.createToken();
}